include/cantera/base/SolutionArray.h

```diff
@@ -0,0 +1,209 @@
+//! @file SolutionArray.h
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef CT_SOLUTIONARRAY_H
+#define CT_SOLUTIONARRAY_H
+
+#include "cantera/base/global.h"
+#include "cantera/base/AnyMap.h"
+
+namespace Cantera
+{
+
+class Solution;
+class ThermoPhase;
+
+/*!
+ *  A container class providing a convenient interface for representing many
+ *  thermodynamic states using the same Solution object. C++ SolutionArray objects are
+ *  one-dimensional by design; extensions to multi-dimensional arrays need to be
+ *  implemented in high-level API's.
+ *
+ *  @since  New in Cantera 3.0.
+ *  @warning This function is an experimental part of the %Cantera API and may be
+ *      changed or removed without notice.
+ */
+class SolutionArray
+{
+private:
+    SolutionArray(const shared_ptr<Solution>& sol,
+                  size_t size,
+                  const AnyMap& meta);
+
+public:
+    virtual ~SolutionArray() {}
+
+    /*!
+     *  Instantiate a new SolutionArray reference
+     *
+     *  @param sol  Solution object defining phase definitions
+     *  @param size  Number of SolutionArray entries
+     *  @param meta  AnyMap holding SolutionArray meta data
+     */
+    static shared_ptr<SolutionArray> create(const shared_ptr<Solution>& sol,
+                                            size_t size=0,
+                                            const AnyMap& meta={})
+    {
+        return shared_ptr<SolutionArray>(new SolutionArray(sol, size, meta));
+    }
+
+    //! Size of SolutionArray (number of entries)
+    int size() const {
+        return m_size;
+    }
+
+    //! SolutionArray meta data.
+    AnyMap& meta() {
+        return m_meta;
+    }
+
+    //! Retrieve associated ThermoPhase object
+    shared_ptr<ThermoPhase> thermo();
+
+    /*!
+     *  Check whether SolutionArray contains a component (property defining state or
+     *  auxiliary variable)
+     */
+    bool hasComponent(const std::string& name) const;
+
+    //! Retrieve a component of the SolutionArray by name
+    vector_fp getComponent(const std::string& name) const;
+
+    /*!
+     *  Set a component of the SolutionArray by name.
+     *
+     *  @param name  Name of component (property defining state or auxiliary variable)
+     *  @param data  Component data
+     *  @param force  If true, add new component to SolutionArray
+     */
+    void setComponent(const std::string& name, const vector_fp& data, bool force=false);
+
+    /*!
+     *  Update the buffered index used to access entries.
+     */
+    void setIndex(size_t index, bool restore=true);
```

Comment:
Since the index is now provided when calling `getState` and `getAuxillary`, I think the `setIndex` method could be made `protected`. The only remaining public use is before calls to `thermo()`, where that method could be modified to have the index as an optional argument (for cases where which state the thermo object is set to don't matter).

Reply:
I agree that within the currently implemented interface there isn't much need for a `public` `setIndex` - the part where it is most useful - picking states within restored `SolutionArray`s from the Python API so that they can be further analyzed - is not implemented at the moment. As such, I'd like to leave the interface as is.